discovery-infra/test_infra/helper_classes/kube_helpers.py

```diff
@@ -0,0 +1,495 @@
+# fixture. With this context manager you will be able to manage a cluster
+# without need to handle registration and deregistration.
+#
+# Example of usage:
+#
+# with cluster_deployment_context(kube_api_client) as cluster_deployment:
+#     print(cluster_deployment.status())
+#
+# When a ClusterDeployment has sufficient data, installation will be started
+# automatically.
+
+import logging
+import os
+import abc
+import contextlib
+
+import waiting
+
+from dataclasses import dataclass
+from pprint import pformat
+from typing import Optional, Union, Dict, Tuple, ContextManager
+
+from kubernetes.config import load_kube_config
+from kubernetes.config.kube_config import Configuration
+from kubernetes.client import ApiClient, CoreV1Api, CustomObjectsApi
+from kubernetes.client.rest import ApiException
+
+from tests.conftest import env_variables
+from test_infra.utils import get_random_name
+
+
+# silence kubernetes debug messages.
+logging.getLogger('kubernetes').setLevel(logging.INFO)
+
+logger = logging.getLogger(__name__)
+
+
+def create_kube_api_client(kubeconfig_path: Optional[str] = None) -> ApiClient:
+    logger.info('creating kube client with config file: %s', kubeconfig_path)
+
+    conf = Configuration()
+    load_kube_config(config_file=kubeconfig_path, client_configuration=conf)
+    return ApiClient(configuration=conf)
+
+
+def _does_string_contain_value(s: Union[str, None]) -> bool:
+    return s and s != '""'
+
+
+@dataclass
+class ObjectReference:
+    """
+    A dataclass that contains the information required to to let you locate a
+    referenced Kube API resource.
+    """
+    name: str
+    namespace: str
+
+    def as_dict(self) -> dict:
+        return {'name': self.name, 'namespace': self.namespace}
+
+
+class Secret:
+    """
+    A Kube API secret resource that consists of the pull secret data, used
+    by a ClusterDeployment CRD.
+    """
+
+    def __init__(
+            self,
+            kube_api_client: ApiClient,
+            name: str,
+            namespace: str = env_variables['namespace']
+    ):
+        self.v1_api = CoreV1Api(kube_api_client)
+        self.__ref = ObjectReference(name=name, namespace=namespace)
+
+    @property
+    def ref(self) -> ObjectReference:
+        return self.__ref
+
+    def create(self, pull_secret: str = env_variables['pull_secret']) -> None:
+        self.v1_api.create_namespaced_secret(
+            body={
+                'apiVersion': 'v1',
+                'kind': 'Secret',
+                'metadata': self.ref.as_dict(),
+                'stringData': {'pullSecret': pull_secret}
+            },
+            namespace=self.ref.namespace
+        )
+
+        logger.info('created secret %s', self.ref)
+
+    def delete(self) -> None:
+        self.v1_api.delete_namespaced_secret(
+            name=self.ref.name,
+            namespace=self.ref.namespace
+        )
+
+        logger.info('deleted secret %s', self.ref)
+
+
+class _BaseCustomResource(abc.ABC):
```

Comment:
underscore in the start of a class name is also not a convention I know
if you want to decide which class is public, use ``__all__``

Reply:
Both ways are acceptable, as I've seen in different places, 
If I'll add `__all__` it will be another (long) list to maintain, I think I prefer to leave that as it

Reply:
``__all__`` doesn't make objects unreachable, it just helps autocompletion in a non-binding way
reconsider